File: fb/103.binary-tree-zigzag-level-order-traversal.cpp
// Time O(n)
// Space O(n)
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(!root) return {};
        vector<vector<int>> ans;
        queue<TreeNode*> q;
        q.push(root);
        bool fromLeft = true;
        while(!q.empty()) {
            int s = q.size();
            vector<int> cur(s, 0);
            for(int i = 0; i < s; i++) {
                TreeNode* n = q.front();
                q.pop();
                if(fromLeft) {
                    cur[i] = n->val;
                }
                else {
                    cur[s - 1 - i] = n->val;
                }
                if(n->left) {
                    q.push(n->left);
                }
                if(n->right) {
                    q.push(n->right);
                }
            }
            ans.push_back(cur);
            if(fromLeft) {
                fromLeft = false;
            }
            else {
                fromLeft = true;
            }
        }
        return ans;
    }
};

File: fb/103.binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []
        if root is None:
            return result
        queue = deque([root])
        tempLevel = deque()
        isFromLeft = True
        
        while len(queue) > 0:
            for i in range(len(queue)):
                node = queue.popleft()
                if isFromLeft:
                    tempLevel.append(node.val)
                else:
                    tempLevel.appendleft(node.val)
                
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                    
            result.append(tempLevel)
            tempLevel = deque()
            isFromLeft = not isFromLeft
                    
        return result
